File: core/src/main/java/com/ctrip/ferriswheel/core/view/Rectangle.java
package com.ctrip.ferriswheel.core.view;

import java.io.Serializable;
import java.util.Objects;

public class Rectangle implements Serializable {
    private int left;
    private int top;
    private int right;
    private int bottom;

    public Rectangle() {
    }

    public Rectangle(int left, int top, int right, int bottom) {
        this.left = left;
        this.top = top;
        this.right = right;
        this.bottom = bottom;
    }

    public int width() {
        return right - left;
    }

    public int height() {
        return bottom - top;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rectangle rectangle = (Rectangle) o;
        return left == rectangle.left &&
                top == rectangle.top &&
                right == rectangle.right &&
                bottom == rectangle.bottom;
    }

    @Override
    public int hashCode() {
        return Objects.hash(left, top, right, bottom);
    }

    public int getLeft() {
        return left;
    }

    public void setLeft(int left) {
        this.left = left;
    }

    public int getTop() {
        return top;
    }

    public void setTop(int top) {
        this.top = top;
    }

    public int getRight() {
        return right;
    }

    public void setRight(int right) {
        this.right = right;
    }

    public int getBottom() {
        return bottom;
    }

    public void setBottom(int bottom) {
        this.bottom = bottom;
    }
}

File: ui/react/src/model/Rectangle.ts
interface Rectangle {
    setRectangle(left: number, top: number, width: number, height: number): void;

    getLeft(): number;

    setLeft(left: number): void;

    getTop(): number;

    setTop(top: number): void;

    getRight(): number;

    getBottom(): number;

    getWidth(): number;

    setWidth(width: number): void

    getHeight(): number;

    setHeight(height: number): void;

    isXInside(x: number): boolean;

    isYInside(y: number): boolean;

    isPointInside(x: number, y: number): boolean;

    moveTo(x: number, y: number): void;

    moveToX(x: number): void;

    moveToY(y: number): void;

};

export default Rectangle;
